### ISSUE TITLE
Fail transactions with invalid cu-limits early

### ISSUE THREAD

**Issue by tao-stones** [anza-xyz/agave#682]
#### Problem

Transaction has a valid CU range: [min_required_cu_for_execution, MAX_COMPUTE_UNIT_LIMIT], where the `min_requested_cu_for_execution` is equal or greater than `Sum(static_native_program_cost)`; meanwhile `compute_budget.set_compute_unit_limit()` takes any value in [0, u64::MAX]. Currently if:
- requested CU > MAX_COMPUTE_UNIT_LIMIT, it is silently capped to `MAX_COMPUTE_UNIT_LIMIT`
- requested CU < `Sum(static_native_program_cost)`, tx fails during execution with `ComputationalBudgetExceeded`

Transaction sets cu limits outside of valid range can be failed early.

#### Proposed Solution
- fail such transactions early at banking_stage with specific error message.

**Comment by apfitzge** [comment]
These are invalid requests:
> 
    requested CU > MAX_COMPUTE_UNIT_LIMIT, it is silently capped to MAX_COMPUTE_UNIT_LIMIT
    requested CU < Sum(static_native_program_cost), tx fails during execution with ComputationalBudgetExceeded


We shouldn't waste resources if we're not going to take fees, and right now we're doing all of this post sigverify.

I don't think we should actually make this change.

**Comment by tao-stones** [comment]
Indeed these two cases makes compute-budget instruction an invalid instruction, so the transaction should be marked with TransactionError as early as possible. Tho this does not mean block producer can't validate payer and collect fee from it.

The proposed change is to correct these:
- for first scenario, the invalid instruction are auto-corrected into valid one, a should-fail transaction is being put through pipeline and being processed instead. 
- for second scenario, transaction goes through pipeline all the way to runtime, where it failed after partial execution when exceeds compute budget. 
In both case, tx can be marked as error then skip execution.

**Comment by apfitzge** [comment]
> Indeed these two cases makes compute-budget instruction an invalid instruction, so the transaction should be marked with TransactionError as early as possible.

Right now those are not instruction failures right? The first case would just silently set the budget to 1.4m but the transaction may not fail. The second will fail.

Agree, we can identify and fast-path any transaction we know will fail by skipping the execution attempt. (This may break RPC reporting?)

> Tho this does not mean block producer can't validate payer and collect fee from it.

I see. Maybe I misunderstood then. I thought you were planning to fail and discard/drop these

**Comment by tao-stones** [comment]
> Right now those are not instruction failures right? The first case would just silently set the budget to 1.4m but the transaction may not fail. The second will fail.

yes, intend is to turn both into instruction error. The first case, imo, is a bug - should not "modify" a transaction at first place;

> I thought you were planning to fail and discard/drop these

Only responsible to fail tx early (and corectly :)), drop or not is block producer decision. (I'm imagining leaders keen to collect fee would happy to fast track such tx to block; and perhaps other leaders care more about what's in block would drop them)

**Comment by apfitzge** [comment]
Great! We're on the same page on this then

**Comment by jstarry** [comment]
How do you calculate priority fees for these failed transactions? If we decide to include the block and collect fees, we need to first agree on how many fees to charge for transactions which set an invalid compute limit because compute limit is used to calculate the priority fee. If someone sets a very high number for the compute limit and we use that value to calculate the priority fee, chances are they won't have enough lamports to pass fee validation anyways.

I think for calculating priority fee we continue capping the compute limit at the max of 1.4M when calculating priority fee. Thoughts?

**Comment by tao-stones** [comment]
It might ultimately be block producer's decision, I honestly think if someone set compute limit above MAX_COMPUTE_UNIT_LIMIT, transaction should just fail without being charged and included. It's like I have a credit card with $1000 limit, should I use it to buy something at $1001, even it is good for credit card issuer to approve this transaction and collect fee, I would expect the issuer to disapprove it and protect myself from mistakes. Unless, of course, my credit card has unlimited spending power; in which cases, we should remove MAX_COMPUTE_UNIT_LIMIT.

**Comment by apfitzge** [comment]
>  I honestly think if someone set compute limit above MAX_COMPUTE_UNIT_LIMIT, transaction should just fail without being charged and included.

I very strongly disagree with this.

> I think for calculating priority fee we continue capping the compute limit at the max of 1.4M when calculating priority fee. 

I think for **fee-calculation** we continue capping at 1.4M, but if the request is invalid its reasonable for us to mark the tx as failed and skip execution.
